libs/contrib/Data/Nat/Factor.idr

```diff
@@ -0,0 +1,332 @@
+module Data.Nat.Factor
+
+import Data.Fin
+import Data.Fin.Extra
+import Data.Nat
+import Decidable.Order
+import Syntax.PreorderReasoning
+
+%default total
+%access public export
+
+
+||| Factor n p is a witness that p is indeed a factor of n,
+||| i.e. there exists a q such that p * q = n.
+data Factor : Nat -> Nat -> Type where
+    CofactorExists : {n, p : Nat} -> (q : Nat) -> n = p * q -> Factor n p
+
+||| NotFactor n p is a witness that p is NOT a factor of n,
+||| i.e. there exist a q and an r, greater than 0 but smaller than p,
+||| such that p * q + r = n.
+data NotFactor : Nat -> Nat -> Type where
+    ProperRemExists : {n, p : Nat} -> (q : Nat) ->
+        (r : Fin (pred p)) ->
+        n = p * q + S (finToNat r) ->
+        NotFactor n p
+
+||| DecFactor n p is a result of the process which decides
+||| whether or not p is a factor on n.
+data DecFactor : Nat -> Nat -> Type where
+    ItIsFactor : Factor n f -> DecFactor n f
+    ItIsNotFactor : NotFactor n f -> DecFactor n f
+
+||| CommonFactor n m p is a witness that p is a factor of both n and m.
+data CommonFactor : Nat -> Nat -> Nat -> Type where
+    CommonFactorExists : {a, b : Nat} -> (p : Nat) -> Factor a p -> Factor b p -> CommonFactor a b p
+
+||| GCD n m p is a witness that p is THE greatest common factor of both n and m.
+||| The second argument to the constructor is a function which for all q being
+||| a factor of both n and m, proves that q is a factor of p.
+|||
+||| This is equivalent to a more straightforward definition, stating that for
+||| all q being a factor of both n and m, q is less than or equal to p, but more
+||| powerful and therefore more useful for further proofs. See below for a proof
+||| that if q is a factor of p then q must be less than or equal to p.
+data GCD : Nat -> Nat -> Nat -> Type where
+    MkGCD : {a, b, p : Nat} ->
+        (CommonFactor a b p) ->
+        ((q : Nat) -> CommonFactor a b q -> Factor p q) ->
+        GCD a b p
+
+
+Uninhabited (Factor (S n) Z) where
+    uninhabited (CofactorExists q prf) = uninhabited prf
+
+Preorder Nat Factor where
+    transitive a b c (CofactorExists qb prfAB) (CofactorExists qc prfBC) =
+        CofactorExists (qb * qc) (
+            rewrite prfAB in
+            rewrite prfBC in
+            rewrite sym $ multAssociative c qc qb in
+            rewrite multCommutative qc qb in
+            Refl
+        )
+
+    reflexive Z = CofactorExists 1 Refl
```

Comment:
This one doesn't require case splitting: `reflexive n = CofactorExists 1 $ sym $ multOneRightNeutral n`

Reply:
Good point